### PR TITLE
Update the contribution guide

### PR DESCRIPTION
**Pull request by freakboy3742**
Update the contribution guide to provide more suggestions for areas to contribute, and to explain how the backend tests work.

Also adds the `{posargs}` improvement to the core test runner.

Lastly, includes a nursery project for a textual backend, since the docs reference that backend as a possible area of contribution.

## PR Checklist:

- [x] All new features have been tested
- [x] All new features have been documented
- [x] I have read the **CONTRIBUTING.md** file
- [x] I will abide by the code of conduct

### CODE REVIEW

File: docs/how-to/contribute-code.rst [hunk 1]
```diff
@@ -354,8 +343,195 @@ Now you are ready to start hacking on Toga!
...
+probably safe to assume that the issue is still available to be worked on.
+
+If an issue is particularly old (more than 6 months), it's entirely possible
+that the issue has been resolved, so the first step is to verify that you can
+reproduce the problem. Use the information provided in the bug report to try and
+reproduce the problem. If you can't reproduce the problem, report what you have
+found as a comment on the ticket, and pick another ticket.
+
+If a bug report has no comments from anyone other that the original reporter,
+the issue needs need to be triaged. Triaging a bug involves taking the
```
Comment (mhsmith) [inline docs/how-to/contribute-code.rst:396]:
```suggestion
If a bug report has no comments from anyone other than the original reporter,
the issue needs to be triaged. Triaging a bug involves taking the
```

File: docs/how-to/contribute-code.rst [hunk 2]
```diff
@@ -354,8 +343,195 @@ Now you are ready to start hacking on Toga!
...
+
+Implement a platform native widget
+----------------------------------
+
+If the core library already specifies an interface for a widget, but the widget
+isn't implemented on your platform of choice, implement that interface. The
+:doc:`supported widgets by platform </reference/widgets_by_platform>` table can
+show you the widgets that are missing on various platforms. You can also look
+for log messages in a running app (or the direct ``factory.not_implemented()``
+function calls that produce those log messages). At present, the web backend is
```
Comment (mhsmith) [inline docs/how-to/contribute-code.rst:461]:
```suggestion
function calls that produce those log messages). At present, the web backend
```

File: docs/how-to/contribute-code.rst [hunk 3]
```diff
@@ -378,19 +554,19 @@ To run the core test suite:
 
       (venv) C:\...>tox -e py-core
 
-You should get some output indicating that tests have been run. You shouldn't
-ever get any FAIL or ERROR test results. We run our full test suite before
-merging every patch. If that process discovers any problems, we don't merge
-the patch. If you do find a test error or failure, either there's something
-odd in your test environment, or you've found an edge case that we haven't
-seen before - either way, let us know!
+You should get some output indicating that tests have been run. You may seek
```
Comment (mhsmith) [inline docs/how-to/contribute-code.rst:557]:
```suggestion
You should get some output indicating that tests have been run. You may see
```

File: docs/how-to/contribute-code.rst [hunk 4]
```diff
@@ -434,11 +607,278 @@ expect to see something like::
...
+
+      (venv) $ tox -e py-core -- tests/path_to_test_file/test_some_test.py
+
+  .. group-tab:: Windows
+
+    .. code-block:: doscon
+
+      (venv) C:\...>tox -e py-core -- tests/path_to_test_file/test_some_test.py
+
+These tests paths are relative to the ``core`` directory. You'll still get a
```
Comment (mhsmith) [inline docs/how-to/contribute-code.rst:635]:
```suggestion
These test paths are relative to the ``core`` directory. You'll still get a
```

File: docs/how-to/contribute-code.rst [hunk 5]
```diff
@@ -434,11 +607,278 @@ expect to see something like::
...
+The widget probe also provides a ``redraw()`` method. GUI libraries don't always
+immediately apply changes visually, as graphical changes will often be batched
+so that they can be applied in a single redraw. To ensure that any visual
+changes have been applied before a test asserts the properties of the app, a
+test case can call ``await probe.redraw()``. This guarantees that any
+outstanding redraw events have been processed. These ``redraw()`` requests are
+also used to implement slow mode - each redraw is turned into a 1 second sleep.
+
+If a widget doesn't have a probe for a given widget, the testbed will
+automatically skip the tests for that widget. If a widget hasn't implemented a
```
Comment (mhsmith) [inline docs/how-to/contribute-code.rst:816]: This isn't true anymore: https://github.com/beeware/toga/pull/1830#issuecomment-1485076558.

File: docs/how-to/contribute-code.rst [hunk 6]
```diff
@@ -466,20 +906,14 @@ submit it without a change note - this will fail, but will also assign a pull
...
 It's not just about coverage!
-=============================
+-----------------------------
 
-Although improving test coverage is the goal, the task ahead of you isn't
-*just* about increasing numerical coverage. Part of the task is to audit the
-code as you go. You could write a comprehensive set of tests for a concrete
-life jacket... but a concrete life jacket would still be useless for the
-purpose it was intended!
+Although we're always trying to improve test coverage is always the goal, the
```
Comment (mhsmith) [inline docs/how-to/contribute-code.rst:912]:
```suggestion
Although we're always trying to improve test coverage, the
```

File: docs/how-to/contribute-docs.rst [hunk 7]
```diff
@@ -116,19 +114,25 @@ documentation:
...
 If a valid spelling of a word is identified as misspelled, then add the word to
 the list in ``docs/spelling_wordlist``. This will add the word to the
-spellchecker's dictionary.
+spellchecker's dictionary. When adding to this list, remember:
+
+* We prefer US spelling, with some liberties for programming-specific
+  colloquialism (e.g., "apps") and verbing of nouns (e.g., "scrollable")
+* Any reference to an product name should use the product's preferred
+  capitalization. (e.g,, "macOS", "GTK", "pytest", "Pygame", "PyScript").
+* If a terms is being used "as code", then it should be quoted as a literal
```
Comment (mhsmith) [inline docs/how-to/contribute-docs.rst:134]:
```suggestion
* Any reference to a product name should use the product's preferred
  capitalization. (e.g., "macOS", "GTK", "pytest", "Pygame", "PyScript").
* If a term is being used "as code", then it should be quoted as a literal
```

File: docs/how-to/contribute-code.rst (hunk 5 above)
Comment (freakboy3742) [reply docs/how-to/contribute-code.rst:816]: Oh yeah - I forgot we did that. Updated.

**Review by mhsmith — APPROVED** [review]
